### PR TITLE
fix: Replace operation queue with simpler implementation.

### PR DESCRIPTION
**Pull request by LBeernaertProton**
Replace the implementation of the user op queue with a simpler
implementation using regular synchronization primitives rather than
channels. The new implementation is much more straight forward and
easier to digest.

`CTQueue` provides an implementation for a closable thread safe queue
and `userOpQueue` takes care of all the merging logic.

Finally, pchan has been removed since it is no longer used.

### CODE REVIEW

File: connector/mock_connector/connector.go [hunk 1]
```diff
@@ -39,15 +39,15 @@ func (m *MockConnector) EXPECT() *MockConnectorMockRecorder {
 // Close mocks base method.
 func (m *MockConnector) Close() error {
```
Comment (jameshoulahan) [inline connector/mock_connector/connector.go:40]: Could you re-generate this? Close method should no longer be here at all...

actually, i think nobody is even using this mock anymore. Maybe just delete the whole `mock_connector` package and `package.go` (which generates it)?

File: internal/backend/user_state.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package backend
 import (
 	"context"
 	"fmt"
+
```
Comment (jameshoulahan) [inline internal/backend/user_state.go:6]: adding this blank line is good but i wonder why there's also a blank line below `gluon/internal/remote`, that must have snuck in somehow. Mind deleting it?

File: internal/parser/proto/imap.pb.go [hunk 3]
```diff
@@ -2020,7 +2020,7 @@ func (x *Close) ProtoReflect() protoreflect.Message {
 	return mi.MessageOf(x)
 }
 
-// Deprecated: Use Close.ProtoReflect.Descriptor instead.
+// Deprecated: Use CloseAndSerializeOperationQueue.ProtoReflect.Descriptor instead.
```
Comment (jameshoulahan) [inline internal/parser/proto/imap.pb.go:2023]: Did you get a bit trigger heavy with the find-replace? :D

File: internal/queue/ctqueue.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+// be closed at any given time. The queue can be closed by either allowing the remaining elements to remain within or
+// to be extracted. Once the queue is closed, no new elements may be added, but existing elements can still be popped of
+// the queue.
+type CTQueue[T any] struct {
+	items  []T
+	cond   *sync.Cond
+	closed int32
+}
+
+func NewTQueue[T any]() *CTQueue[T] {
```
Comment (jameshoulahan) [inline internal/queue/ctqueue.go:18]:
```suggestion
func NewCTQueue[T any]() *CTQueue[T] {
```

File: internal/queue/ctqueue.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	var item T
+	item, ctq.items = ctq.items[0], ctq.items[1:]
+
+	return item, true
+}
+
+// TryPop attempts to pop an element of the queue, but if no elements are available it doesn't wait and exists
+//immediately.
```
Comment (jameshoulahan) [inline internal/queue/ctqueue.go:63]:
```suggestion
// TryPop attempts to pop an element of the queue, but if no elements are available it doesn't wait and exits
// immediately.
```

File: internal/queue/ctqueue.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+)
+
+// CTQueue is a thread safe queue implementation which allows safe usage among different threads/go-routines and can
+// be closed at any given time. The queue can be closed by either allowing the remaining elements to remain within or
+// to be extracted. Once the queue is closed, no new elements may be added, but existing elements can still be popped of
+// the queue.
+type CTQueue[T any] struct {
+	items  []T
+	cond   *sync.Cond
+	closed int32
```
Comment (jameshoulahan) [inline internal/queue/ctqueue.go:15]: Go 1.19 beta includes an atomic bool type. Let's keep this as an int32 for now but remember to swap it for an atomic bool when 1.19 goes to stable (or at least release candidate, such that it's available for download on the go homepage).

cf https://pkg.go.dev/sync/atomic@master#Bool

File: internal/queue/ctqueue.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// CTQueue is a thread safe queue implementation which allows safe usage among different threads/go-routines and can
+// be closed at any given time. The queue can be closed by either allowing the remaining elements to remain within or
+// to be extracted. Once the queue is closed, no new elements may be added, but existing elements can still be popped of
+// the queue.
+type CTQueue[T any] struct {
```
Comment (jameshoulahan) [inline internal/queue/ctqueue.go:12]: PChan had some unit tests which caught a few issues in the past. I'd feel more comfortable if this type had some unit tests as well.

File: internal/remote/mock_remote/remote.go [hunk 8]
```diff
@@ -39,15 +39,15 @@ func (m *MockConnector) EXPECT() *MockConnectorMockRecorder {
 // Close mocks base method.
 func (m *MockConnector) Close() error {
 	m.ctrl.T.Helper()
-	ret := m.ctrl.Call(m, "Close")
+	ret := m.ctrl.Call(m, "CloseAndSerializeOperationQueue")
```
Comment (jameshoulahan) [inline internal/remote/mock_remote/remote.go:42]: More find-replace collateral damage?

File: internal/remote/user.go [hunk 9]
```diff
@@ -5,13 +5,13 @@ import (
 	"encoding/gob"
 	"errors"
 	"fmt"
+	"github.com/ProtonMail/gluon/internal/queue"
```
Comment (jameshoulahan) [inline internal/remote/user.go:8]: Could you run goimports on this file? Import ordering is a bit off.

File: internal/remote/user_queue.go [hunk 10]
```diff
@@ -3,109 +3,152 @@ package remote
...
-// TODO: What should we do with operations that failed to execute due to auth reasons?
-// We might want to save them somewhere so we can try again after the user has logged back in.
-func (user *User) process() {
-	defer user.processWG.Done()
+type userOpQueue struct {
+	poppedOps []operation
+	queue     *queue.CTQueue[operation]
+}
+
+func newUserOpQueue(queue *queue.CTQueue[operation]) userOpQueue {
```
Comment (jameshoulahan) [inline internal/remote/user_queue.go:17]: Why pass the queue in here DI-style? Any reason not to just construct a new queue internally?

File: internal/remote/user.go [hunk 11]
```diff
@@ -52,7 +48,7 @@ func newUser(userID, path string, conn connector.Connector) (*User, error) {
 		path:              path,
 		conn:              conn,
 		updatesCh:         make(chan imap.Update),
-		queue:             pchan.New[operation](),
+		mergeQueue:        newUserOpQueue(queue.NewTQueue[operation]()),
```
Comment (jameshoulahan) [inline internal/remote/user.go:51]: why is it called mergeQueue? why not just simply `opQueue`?

File: internal/backend/user_state.go (hunk 2 above)
Comment (LBeernaertProton) [reply internal/backend/user_state.go:6]: goimports adds a new line here.

Comment (jameshoulahan) [reply internal/backend/user_state.go:6]: Yes, as it should. I'm referring to the other blank line:

```go
	"github.com/ProtonMail/gluon/internal/remote"
// this one snuck in somehow... it should be removed
	"github.com/ProtonMail/gluon/internal/backend/ent"
```

File: internal/queue/ctqueue.go (hunk 4 above)
Comment (LBeernaertProton) [reply internal/queue/ctqueue.go]: Fixed.

File: internal/parser/proto/imap.pb.go (hunk 3 above)
Comment (LBeernaertProton) [reply internal/parser/proto/imap.pb.go]: Yep. Fixed.

File: connector/mock_connector/connector.go (hunk 1 above)
Comment (LBeernaertProton) [reply connector/mock_connector/connector.go]: Done

File: internal/queue/ctqueue.go (hunk 5 above)
Comment (LBeernaertProton) [reply internal/queue/ctqueue.go]: Good catch. Fixed.

File: internal/remote/mock_remote/remote.go (hunk 8 above)
Comment (LBeernaertProton) [reply internal/remote/mock_remote/remote.go:42]: Fixed.

File: internal/remote/user.go (hunk 9 above)
Comment (LBeernaertProton) [reply internal/remote/user.go]: Done.

File: internal/remote/user_queue.go (hunk 10 above)
Comment (LBeernaertProton) [reply internal/remote/user_queue.go]: Good point. It was a left over from a previous state where that wasn't possible. Fixed.

File: internal/remote/user.go (hunk 11 above)
Comment (LBeernaertProton) [reply internal/remote/user.go]: Done.

File: go.sum [hunk 12]
```diff
@@ -60,10 +60,12 @@ github.com/kr/text v0.2.0 h1:5Nx0Ya0ZqY2ygV366QzturHI13Jq95ApcVaJBhpS+AY=
 github.com/kr/text v0.2.0/go.mod h1:eLer722TekiGuMkidMxC/pM04lWEeraHUUmBw8l2grE=
 github.com/kylelemons/godebug v0.0.0-20170820004349-d65d576e9348 h1:MtvEpTB6LX3vkb4ax0b5D2DHbNAUsen0Gx5wZoq3lV4=
 github.com/kylelemons/godebug v0.0.0-20170820004349-d65d576e9348/go.mod h1:B69LEHPfb2qLo0BaaOLcbitczOKLWTsrBG9LczfCD4k=
+github.com/mattn/go-runewidth v0.0.9 h1:Lm995f3rfxdpd6TSmuVCHVb/QhupuXlYr8sCI/QdE+0=
```
Comment (jameshoulahan) [inline go.sum:63]: don't forget `go mod tidy` after code generation -- the ent generation step leaves behind unnecessary go.sum entries.

Also i suggest running goimports on the whole codebase after generation too; the swig generated code has weird tab widths and other oddities

File: internal/backend/user_state.go (hunk 2 above)
Comment (LBeernaertProton) [reply internal/backend/user_state.go:6]: Fixed

File: internal/queue/ctqueue.go (hunk 7 above)
Comment (LBeernaertProton) [reply internal/queue/ctqueue.go:12]: Done

File: go.sum (hunk 12 above)
Comment (LBeernaertProton) [reply go.sum]: Done

File: internal/queue/ctqueue_test.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+	require.Zero(t, queue.Len())
+}
+
+func TestCTQueue_CloseAndPop(t *testing.T) {
+	queue := NewCTQueue[int]()
+	expectedValues := []int{20, 30, 40, 50}
+
+	go func() {
+		// Sleep for 0.5 sec
+		time.Sleep(500000000)
```
Comment (jameshoulahan) [inline internal/queue/ctqueue_test.go:50]:
```suggestion
		time.Sleep(500 * time.Millisecond)
```

File: internal/queue/ctqueue_test.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+	require.True(t, queue.IsClosed())
+	require.Zero(t, queue.Len())
+}
+
+func TestCTQueue_CloseWakesBlockedPop(t *testing.T) {
+	queue := NewCTQueue[int]()
+
+	go func() {
+		// Sleep for 0.5 sec
+		time.Sleep(500000000)
```
Comment (jameshoulahan) [inline internal/queue/ctqueue_test.go:78]:
```suggestion
		time.Sleep(500 * time.Millisecond)
```

**Review by jameshoulahan — CHANGES_REQUESTED** [review]

File: internal/queue/ctqueue_test.go (hunk 13 above)
Comment (LBeernaertProton) [reply internal/queue/ctqueue_test.go:50]: Done

File: internal/queue/ctqueue_test.go (hunk 14 above)
Comment (LBeernaertProton) [reply internal/queue/ctqueue_test.go:78]: Done

**Pull request merged**